lightning/src/ln/peers/outbound_queue.rs

```diff
@@ -0,0 +1,350 @@
+/// Abstracts the buffer used to write data through a SocketDescriptor handling partial writes and
+/// flow control.
+
+use ln::peers::handler::{SocketDescriptor, PayloadQueuer, SocketDescriptorFlusher};
+use std::collections::LinkedList;
+use std::cmp;
+
+pub(super) struct OutboundQueue {
+	blocked: bool,
+	soft_limit: usize,
+	buffer: LinkedList<Vec<u8>>,
+	buffer_first_msg_offset: usize,
+}
+
+impl PayloadQueuer for OutboundQueue {
+	/// Unconditionally queue item. May increase queue above soft limit.
+	fn push_back(&mut self, item: Vec<u8>) {
```

Comment:
None of this code has behavior changes, but I'm glad my comments and refactoring made it more clear what the current code does.

I completely agree that this is the wrong approach, but this PR isn't about changing the behavior it is about making the code modular, testable, and fixing bugs. Then it will be safe to change.

The story in the backlog (which I'm in the process of turning into Github Issues) has a solution that involves dropping optional messages. I'll leave that to a design discussion outside this patch.

Reply:
I changed the API to `flush()` and it flushes the entire queue. It is in a separate PR found here: https://github.com/rust-bitcoin/rust-lightning/pull/714

Unfortunately, this had to go as a separate PR at the end as it encountered an existing bug (5c4cda7) fixed in 3/3. Luckily, it also found 2 more exiting bugs so the quality bar is going up. Let me know what you think.